numpy/core/arrayprint.py

```diff
@@ -607,14 +600,23 @@ def _formatArray(a, format_function, rank, max_line_len,
 
 class FloatFormat(object):
     def __init__(self, data, precision, suppress_small, sign=False):
+        # for backcompatibility, accept bools
+        if sign is False:
+            sign = '-'
+        if sign is True:
+            sign = '+'
+
+        # for backcompatibility, 0d arrays are not padded
+        if sign == ' ' and data.shape == ():
```

Comment:
Why do we do `data.shape == ()` here, but `data.size == 0` in `BoolFormat`. Let's just use the latter in both places

Reply:
Ah, there is the answer to my question at the top - I guess the real question is whether we want this back-compatibility. I'd argue no, but only weakly. Maybe an alternative is to have a special `sign` for backward compatibility? E.g., have this be the meaning of `sign=False`?

Reply:
The motivation was this: If anybody is really unhappy with our whitespace changes, they can get back to virtually the same behavior as in 1.13 by doing `np.set_printoptions(sign=' ')` in their test suite. The 0d special case reproduces 1.13 behavior.

This way they can easily shut up their doctests until they have time to properly fix them.

Reply:
I suppose we could also add an extra sign option `sign='legacy'` or similar.

Reply:
@ahaldane: you're responding to mhvk and not me here, right? What are your thoughts on using `data.size == 0` here instead?

Reply:
I like the `sign='legacy'` option - that way, space can be completely consistent with the standard python format specification.

Reply:
@mhvk, OK, will add the `sign='legacy'` option. I am a little busy this week so there may be a short delay.

@eric-wieser I wanted to use `data.shape == ()` here to catch only 0-d arrays, and `data.size > 0` in `BoolFormat` to catch both 0d and nd-arrays-of-size-1, eg `np.array([4.])`. The goal was to preserve 0d array behavior in general, but also remove the space in size-1 Bool arrays. Note that 0d arrays have `data.size = 1`

Reply:
Why do we want `array([True])` yet `array([ 1.0])` to be the defaults?

It seems like we should be consistent here, and choose either:

*  `size == 1`: `array([True])` and `array([1.0])
* `shape == ()`: `array([ True])` and `array([ 1.0])`

I'd favor the first one, because it lets us keep the model that the formatter gets given the flattened array, rather than _"it's 1D except when it isn't"_

Reply:
In my mind, the space in bool arays is not a "sign" character, it is a space for alignment. Bools don't have a sign.  

This is different from the space in the float array, which represents the sign position. I think we need to keep in in size-1 arrays, because the space should only appear now if the user specified `sign=' '` (which is not the default), meaning that they explicitly said they wanted a space in the sign position. 

Also, `array([1.0])` *is* the default, since `sign='-'` is the new default in this PR.